server/lomas_server/dp_queries/dp_libraries/smartnoise_synth.py

```diff
@@ -0,0 +1,307 @@
+            #     constraints[col] = ChainTransformer(
+            #         [
+            #             DateTimeTransformer(),
+            #             BinTransformer(bins=20, nullable=nullable),
+            #         ]
+            #     )
+            for col in col_categories[SSynthColumnType.ORDINAL]:
+                constraints[col] = LabelTransformer(nullable=nullable)
+
+        return TableTransformer.create(
+            data=private_data,
+            style=style,
+            nullable=nullable,
+            constraints=constraints,
+        )
+
+    def _preprocess_data(
+        self,
+        private_data: pd.DataFrame,
+        query_json: dict,
+    ) -> pd.DataFrame:
+        """
+        Preprocess the data based on the query parameters.
+
+        Args:
+            private_data (pd.DataFrame): Private data to be preprocessed
+            query_json (dict): (SmartnoiseSynthModelCost): JSON request object
+                select_cols (List[str]): List of columns to select
+                mul_matrix (List): Multiplication matrix for columns aggregations
+
+        Returns:
+            pd.DataFrame: Preprocessed private data
+        """
+        if query_json.select_cols:
+            try:
+                private_data = private_data[query_json.select_cols]
+            except KeyError as e:
+                raise InvalidQueryException(
+                    "Error while selecting provided select_cols: " + str(e)
+                ) from e
+
+        if query_json.mul_matrix:
+            try:
+                np_matrix = np.array(query_json.mul_matrix)
+                mul_private_data = private_data.to_numpy().dot(np_matrix.T)
+                private_data = pd.DataFrame(mul_private_data)
+            except ValueError as e:
+                raise InvalidQueryException(
+                    f"Failed to multiply provided mul_matrix: {(str(e))}"
+                ) from e
+        return private_data
+
+    def _get_fit_model(
+        self,
+        private_data: pd.DataFrame,
+        transformer: TableTransformer,
+        query_json: dict,
+    ) -> Synthesizer:
+        """
+        Create and fit the synthesizer model.
+
+        Args:
+            private_data (pd.DataFrame): Private data for fitting the model
+            transformer (TableTransformer): Transformer to pre/postprocess data
+            query_json (SmartnoiseSynthModelCost): JSON request object for the query
+                synth_name (str): name of the Yanthesizer model to use
+                epsilon (float): epsilon budget value
+                delta (float): delta budget value
+                nullable (bool): True if some data cells may be null
+                synth_params (dict): Keyword arguments to pass to the synthesizer
+                    constructor.
+
+        Returns:
+            Synthesizer: Fitted synthesizer model
+        """
+        if query_json.delta:  # not all model take delta as argument
```

Comment:
Do we want to check for this to warn the user or does the synthesizer.create method throw an exception?